src/cache/cache_las.c

```diff
@@ -859,20 +884,66 @@ __wt_las_sweep(WT_SESSION_IMPL *session)
 		 * should another thread remove the record before we do (not
 		 * expected for dropped trees), and the cursor remains
 		 * positioned in that case.
-		 *
-		 * TODO it would also be good to remove entries in lookaside
-		 * from live files that have aged out.  If we track for each
-		 * entry whether it was the on-page value chosen by
-		 * reconciliation, we can safely remove entries from that point
-		 * on (for the given key) that are visible to all readers.
 		 */
 		if (las_id >= cache->las_sweep_dropmin &&
 		    las_id <= cache->las_sweep_dropmax &&
 		    __bit_test(cache->las_sweep_dropmap,
 		    las_id - cache->las_sweep_dropmin)) {
 			WT_ERR(cursor->remove(cursor));
 			++remove_cnt;
+			saved_key.data = NULL;
+			continue;
 		}
+
+		/*
+		 * Remove entries from the lookaside that have aged out and are
+		 * now no longer needed.
+		 */
+		WT_ERR(cursor->get_value(cursor,
+		    &txnid, &las_timestamp, &upd_type, &las_value));
+		val_ts = NULL;
+#ifdef HAVE_TIMESTAMPS
+		WT_ASSERT(session, las_timestamp.size == WT_TIMESTAMP_SIZE);
+		memcpy(&timestamp, las_timestamp.data, las_timestamp.size);
+		val_ts = &timestamp;
+#endif
+
+		/*
+		 * Save our key for comparing with older entries if we
+		 * don't have one or it is different.
+		 */
+		if (saved_key.data == NULL ||
+		    saved_key.size != las_key.size ||
+		    memcmp(saved_key.data,
+		    las_key.data, las_key.size) != 0) {
+			WT_ERR(__wt_cursor_get_raw_key(
+			    cursor, &saved_key));
+			if (!WT_DATA_IN_ITEM(&saved_key))
+				WT_ERR(__wt_buf_set(session, &saved_key,
+				    saved_key.data, saved_key.size));
+			newkey = true;
+		} else
+			newkey = false;
+
+		/*
+		 * If this entry isn't globally visible we cannot remove it.
+		 * If it is visible then perform additional checks to see
+		 * whether it has aged out of a live file.
+		 */
+		if (!__wt_txn_visible_all(session, txnid, val_ts))
+			continue;
```

Comment:
Did we decide the `__wt_txn_visible_all` call isn't expensive enough to worry about? Or is there a case where one visible entry doesn't imply subsequent entries can be discarded?

Reply:
I'm not worried about it: I haven't seen evidence of it in perf traces.